arroyo/processing/strategies/unfold.py

```diff
@@ -0,0 +1,104 @@
+import time
+from collections import deque
+from typing import Callable, Collection, Deque, Generic, Optional, TypeVar, Union, cast
+
+from arroyo.processing.strategies.abstract import MessageRejected, ProcessingStrategy
+from arroyo.types import FilteredPayload, Message, Value
+
+TInput = TypeVar("TInput")
+TOutput = TypeVar("TOutput")
+
+
+class Unfold(
+    ProcessingStrategy[Union[FilteredPayload, TInput]], Generic[TInput, TOutput]
+):
+    """
+    Unfold receives a message and explodes it to generate a collection of
+    messages submitting them one by one to the next step. The generated
+    messages are created according to the generator function provided by the user.
+
+    The generator function provided must return a collection (i.e. a class that
+    implements sized + iterable).
+
+    If this step receives a `MessageRejected` exception from the next
+    step it keeps the remaining messages and attempts to submit
+    them on subsequent calls to `poll`
+    """
+
+    def __init__(
+        self,
+        generator: Callable[[TInput], Collection[TOutput]],
+        next_step: ProcessingStrategy[Union[FilteredPayload, TOutput]],
+    ) -> None:
+        self.__generator = generator
+        self.__next_step = next_step
+        self.__closed = False
+        # If we get MessageRejected from the next step, we put the pending messages here
+        self.__pending: Deque[Message[TOutput]] = deque()
+
+    def submit(self, message: Message[Union[FilteredPayload, TInput]]) -> None:
+        assert not self.__closed
+        if self.__pending:
+            raise MessageRejected
+
+        if isinstance(message.payload, FilteredPayload):
+            self.__next_step.submit(
```

Comment:
this reorders messages, so you can't do this. imagine three messages where the second one is filtered. it will be sorted in front of the other ones

Reply:
oh right...